File: haskell/src/Model/Skill.hs
module Model.Skill where

import Data.Aeson (ToJSON(..), object, (.=), Value(Null))
import Data.Maybe (catMaybes)
import Data.Text (Text)
import GHC.Generics (Generic)

data SkillItem = SkillItem
  { skName         :: Text
  , skCategory     :: Text
  , skYears        :: Text
  , skProficiency  :: Maybe Text
  , skPicture      :: Maybe Text
  , skPictureColor :: Maybe Text
  } deriving (Show, Generic)

instance ToJSON SkillItem where
  toJSON sk = object $ baseFields ++ optionalFields
    where
      baseFields =
        [ "name"     .= skName sk
        , "category" .= skCategory sk
        , "years"    .= skYears sk
        ]
      optionalFields = catMaybes
        [ fmap (\v -> "proficiency"  .= v) (skProficiency sk)
        , fmap (\v -> "picture"      .= v) (skPicture sk)
        , fmap (\v -> "pictureColor" .= v) (skPictureColor sk)
        ]

File: go/model/skill.go
package model

// SkillItem represents a single skill entry.
type SkillItem struct {
	Name         string `json:"name"`
	Category     string `json:"category"`
	Years        string `json:"years"`
	Proficiency  string `json:"proficiency,omitempty"`
	Picture      string `json:"picture,omitempty"`
	PictureColor string `json:"pictureColor,omitempty"`
}
